udp/writer.go

```diff
@@ -0,0 +1,63 @@
+// Copyright 2015 The Chihaya Authors. All rights reserved.
+// Use of this source code is governed by the BSD 2-Clause license,
+// which can be found in the LICENSE file.
+
+package udp
+
+import (
+	"bytes"
+	"encoding/binary"
+	"time"
+
+	"github.com/chihaya/chihaya/tracker/models"
+)
+
+// Writer implements the tracker.Writer interface for the UDP protocol.
+type Writer struct {
+	buf *bytes.Buffer
+
+	connectionID  []byte
+	transactionID []byte
+}
+
+// WriteError writes the failure reason as a null-terminated string.
+func (w *Writer) WriteError(err error) error {
+	w.writeHeader(errorActionID)
+	w.buf.WriteString(err.Error())
+	w.buf.WriteRune('\000')
+	return nil
+}
+
+// WriteAnnounce encodes an announce response according to the UDP spec.
+func (w *Writer) WriteAnnounce(res *models.AnnounceResponse) error {
+	w.writeHeader(announceActionID)
```

Comment:
I'm not sure if other projects refer to it as a header, but it still is a header that all the responses have. I don't particularly care either way though.